README.rst
.. image:: https://travis-ci.org/matthew-brett/wheels2dmg.svg?branch=master
    :target: https://travis-ci.org/matthew-brett/wheels2dmg

##########
wheels2dmg
##########

A Python utility to build an OSX disk image and ``.pkg`` installer to install
a given set of wheels.

The idea is that any package that can be installed via pip can also have a
safe dmg installer, because we use pip to install the wheels, and pip will
manage the dependencies.

The installer only installs into Python.org Python, but includes instructions
on how to install manually into macports and homebrew.

Here's how to make an installer for matplotlib 1.4.0::

    wheels2dmg matplotlib 1.4.0 matplotlib --python-version 3.4.1

Or for more than one package::

    wheels2dmg scipy-stack 1.0 scipy matplotlib ipython[notebook] \
        pandas numexpr sympy --python-version 2.7.8

.. warning::

    Please be careful - this software is alpha quality and has not been much
    tested in the wild.  I'd love to get feedback if you have problems, and
    I'd be very happy to have pull requests to this repo too.

****
Code
****

See https://github.com/matthew-brett/wheels2dmg

Released under the BSD two-clause license - see the file ``LICENSE`` in the
source distribution.

`travis-ci <https://travis-ci.org/matthew-brett/wheels2dmg>`_ kindly tests the
code automatically under Python 2.7, 3.3 and 3.4.

The latest released version is at https://pypi.python.org/pypi/wheels2dmg

*******
Support
*******

Please put up issues on the `wheels2dmg issue tracker
<https://github.com/matthew-brett/wheels2dmg/issues>`_.


Changelog
.. -*- mode: rst -*-
.. vim:ft=rst

.. _changelog:

#####################
Wheels2dmg change log
#####################

********
Releases
********

* 0.2 (Sunday October 11 2014)

  Second public release

  * Added travis-ci testing
  * Make wheel delocation optional and off by default

* 0.1

  First public release


wheels2dmg/tests/test_piputils.py
""" Testing piputils module
"""

from os.path import join as pjoin, abspath, dirname

from pip.exceptions import InstallationError

from ..piputils import (make_pip_parser, recon_pip_args, get_requirements,
                        get_req_strings)
from ..wheels2dmg_cmd import get_parser
from ..tmpdirs import InTemporaryDirectory

from nose.tools import (assert_true, assert_false, assert_raises,
                        assert_equal, assert_not_equal)


DATA_PATH = pjoin(abspath(dirname(__file__)), 'data')

SCIPY_STACK_NAMES = [
    'pip>=1.5.4',
    'numpy>=1.6',
    'scipy>=0.10',
    'pytz',
    'tornado',
    'python-dateutil',
    'matplotlib>=1.1',
    'pyzmq',
    'ipython[notebook,test]>=0.13',
    'numexpr',
    'pandas>=0.8',
    'sympy>=0.7',
    'nose>=1.1']

SCIPY_STATS_DATA_NAMES = SCIPY_STACK_NAMES + [
    'scikit-learn',
    'scikit-image',
    'patsy',
    'statsmodels',
    'h5py',
    'Pillow']


def test_reconstruct_pip_args():
    # Function to reconstruct pip command line arguments
    class A():
        req_specs = None
        requirement = None
        index_url = None
        extra_index_url = None
        no_index = None
        find_links = None
    args = A()
    assert_equal(recon_pip_args(args), ([], []))
    args.req_specs = ()
    assert_equal(recon_pip_args(args), ([], []))
    args.req_specs = ['foot', 'barrel']
    exp_reqs = ['foot', 'barrel']
    assert_equal(recon_pip_args(args), (exp_reqs, []))
    args.requirement = ()
    assert_equal(recon_pip_args(args), (exp_reqs, []))
    args.requirement = ('file1.txt', 'file2.txt')
    exp_reqs += ['--requirement=file1.txt', '--requirement=file2.txt']
    assert_equal(recon_pip_args(args), (exp_reqs, []))
    args.index_url = 'http://matthew.dynevor.org'
    exp_others = ['--index-url=http://matthew.dynevor.org']
    assert_equal(recon_pip_args(args), (exp_reqs, exp_others))
    args.extra_index_url = ()
    assert_equal(recon_pip_args(args), (exp_reqs, exp_others))
    extra_urls = ('http://www.python.org', 'http://wheels.scikit-image.org')
    args.extra_index_url = extra_urls
    exp_others += ['--extra-index-url=' + u for u in extra_urls]
    assert_equal(recon_pip_args(args), (exp_reqs, exp_others))
    # False does not trigger --no-index
    args.no_index = False
    assert_equal(recon_pip_args(args), (exp_reqs, exp_others))
    args.no_index = True
    exp_others += ['--no-index']
    assert_equal(recon_pip_args(args), (exp_reqs, exp_others))
    args.find_links = ()
    assert_equal(recon_pip_args(args), (exp_reqs, exp_others))
    link_urls = ('http://nipy.org', 'http://nipy.org/nibabel')
    args.find_links = link_urls
    exp_others += ['--find-links=' + u for u in link_urls]
    assert_equal(recon_pip_args(args), (exp_reqs, exp_others))


def test_parser_reconstruct_pip():
    # Check roundtrip using our parser and recon_pip_args
    for parser, null_args in (
        (get_parser(), ['pkg_name', '1.0']),
        (make_pip_parser(), [])
    ):
        for in_out_args in [
            # arguments to parser, [expected arguments to send to pip]
            # If second element missing assume input + output same
            ([],),
            (['--requirement=file1.txt'],),
            (['-r', 'file1.txt'], ['--requirement=file1.txt']),
            (['--requirement=file1.txt', '--requirement=file2.txt'],),
            (['-r', 'file1.txt', '-r', 'file2.txt'],
            ['--requirement=file1.txt', '--requirement=file2.txt']),
            (['--index-url=http://www.foo.com'],),
            (['-i', 'http://www.foo.com'], ['--index-url=http://www.foo.com']),
            (['--extra-index-url=http://www.foo.com'],),
            (['--extra-index-url=http://www.foo.com'],),
            (['--extra-index-url=http://www.foo.com',
            '--extra-index-url=http://bar.org'],),
            (['--no-index'],),
            (['--find-links=http://www.foo.com'],),
            (['--find-links=http://www.foo.com',
            '--find-links=http://bar.org'],),
            (['-f', 'http://www.foo.com', '--find-links=http://bar.org'],
            ['--find-links=http://www.foo.com',
            '--find-links=http://bar.org'],),
            (['foo', 'bar'],)
        ]:
            if len(in_out_args) == 1:
                in_args = in_out_args[0]
                out_args = in_args
            else:
                in_args, out_args = in_out_args
            parsed = parser.parse_args(null_args + in_args)
            assert_equal(sum(recon_pip_args(parsed), []), out_args)


def assert_names_equal(req_set, names, *args, **kwargs):
    assert_equal(names, get_req_strings(req_set, *args, **kwargs))


def test_get_requirements():
    # Test get_requirements function
    # Gets set of pip requiremnts from input arguments
    assert_names_equal(get_requirements([]), [])
    # Can use None for default second argument
    assert_names_equal(get_requirements([], None), [])
    assert_names_equal(get_requirements(['one']), ['one'])
    assert_names_equal(get_requirements(['one', 'two==1.2']),
                       ['one', 'two==1.2'])
    assert_names_equal(get_requirements(['one[an_extra]', 'two==1.2']),
                       ['one[an_extra]', 'two==1.2'])
    assert_names_equal(
        get_requirements(['one[an_extra,another]', 'two==1.2']),
                       ['one[an_extra,another]', 'two==1.2'])
    assert_names_equal(get_requirements(['one', 'two==1.2,<1.3']),
                       ['one', 'two==1.2,<1.3'])
    sps_file = pjoin(DATA_PATH, 'scipy-stack-1.0-plus.txt')
    assert_names_equal(get_requirements([], [sps_file]), SCIPY_STACK_NAMES)
    assert_names_equal(get_requirements(['three', 'four==5'], [sps_file]), 
                       ['three', 'four==5'] + SCIPY_STACK_NAMES)
    # Duplicate requirements cause an error
    assert_raises(InstallationError, get_requirements, ['three', 'three'])
    assert_raises(InstallationError, get_requirements, ['three', 'three==5.0'])
    assert_raises(InstallationError, get_requirements, ['numpy'], [sps_file])
    # A requirements file embedding a requirements file
    with InTemporaryDirectory():
        with open('more_required.txt', 'wt') as fobj:
            fobj.write("""
# A comment
# An unused URL
-f https://nipy.bic.berkeley.edu/scipy_installers
-r {0}
scikit-learn
scikit-image
patsy
statsmodels
h5py
Pillow
""".format(sps_file))
        assert_names_equal(get_requirements([], ['more_required.txt']),
                            SCIPY_STATS_DATA_NAMES)


def test_get_req_strings():
    # Test extras and versions parameters to get_req_strings
    assert_names_equal(get_requirements(['one']), ['one'])
    assert_names_equal(get_requirements(['one']), ['one'], False, False)
    assert_names_equal(get_requirements(['one', 'two==1.2']),
                       ['one', 'two==1.2'], False)
    assert_names_equal(get_requirements(['one', 'two==1.2']),
                       ['one', 'two'], False, False)
    assert_names_equal(get_requirements(['one[an_extra]', 'two==1.2']),
                       ['one', 'two==1.2'], False, True)
    assert_names_equal(get_requirements(['one[an_extra]', 'two==1.2']),
                       ['one[an_extra]', 'two'], True, False)
    assert_names_equal(get_requirements(['one[an_extra]', 'two==1.2']),
                       ['one', 'two'], False, False)


wheels2dmg/tests/test_pkgbuilders.py
""" Testing pkgbuilders module
"""

from os.path import (basename, dirname, abspath, expanduser, relpath,
                     join as pjoin)

from ..pkgbuilders import (get_get_pip, insert_template_path,
                           pop_template_path, get_template,
                           PkgWriter)

from ..tmpdirs import TemporaryDirectory

from nose.tools import (assert_true, assert_false, assert_raises,
                        assert_equal, assert_not_equal)

TEMPLATE_PATH = abspath(pjoin(dirname(__file__), '..', 'templates'))


def assert_file_equal(file1, file2):
    with open(file1, 'rb') as fobj:
        contents1 = fobj.read()
    with open(file2, 'rb') as fobj:
        contents2 = fobj.read()
    assert_equal(contents1, contents2)


def assert_file_equal_string(fname, text):
    with open(fname, 'rt') as fobj:
        contents = fobj.read()
    assert_equal(contents, text)


def test_get_get_pip():
    # Test get_get_pip function
    with TemporaryDirectory() as tmpdir:
        gpp = get_get_pip(__file__, tmpdir)
        assert_equal(basename(gpp), 'get-pip.py')
        assert_file_equal(__file__, gpp)
    with TemporaryDirectory() as tmpdir:
        url = 'file://' + abspath(__file__)
        gpp = get_get_pip(url, tmpdir)
        assert_file_equal(__file__, gpp)
    # Check get_pip file can contain tilde
    home = expanduser('~')
    me_homed = pjoin('~', relpath(__file__, home))
    with TemporaryDirectory() as tmpdir:
        gpp = get_get_pip(me_homed, tmpdir)
        assert_file_equal(__file__, gpp)


def test_py_version_strings():
    pkg_writer = PkgWriter('test', '1', '2.7.1', ['foo', 'bar'])
    assert_equal(pkg_writer.pyv_m_m_e, '2.7.1')
    assert_equal(pkg_writer.pyv_m_m, '2.7')
    assert_equal(pkg_writer.pyv_mm, '27')
    assert_equal(pkg_writer.pyv_m, '2')
    for ver in ('2', '2.', '2.7', '2.7.'):
        assert_raises(ValueError, PkgWriter, 'test', '1', ver, ['foo', 'bar'])


def test_get_requirement_strings():
    # Test extras and versions parameters to get_req_strings
    def assert_pkg_reqs_equal(pip_args, expected, *args, **kwargs):
        pkg_writer = PkgWriter('test', '1', '2.7.1', pip_args)
        reqs = pkg_writer.get_requirement_strings(*args, **kwargs)
        assert_equal(reqs, expected)

    assert_pkg_reqs_equal(['one'], ['one'])
    assert_pkg_reqs_equal(['one'], ['one'], False, False)
    assert_pkg_reqs_equal(['one', 'two==1.2'], ['one', 'two==1.2'], False)
    assert_pkg_reqs_equal(['one', 'two==1.2'], ['one', 'two'], False, False)
    assert_pkg_reqs_equal(['one[an_extra]', 'two==1.2'],
                          ['one', 'two==1.2'], False, True)
    assert_pkg_reqs_equal(['one[an_extra]', 'two==1.2'],
                          ['one[an_extra]', 'two'], True, False)
    assert_pkg_reqs_equal(['one[an_extra]', 'two==1.2'],
                          ['one', 'two'], False, False)


def test_write_requires():
    # Test write_require function
    pkg_writer = PkgWriter('test', '1', '2.7.1', ['foo', 'bar'])
    exp_out = pjoin(pkg_writer.wheel_build_dir, 'test-1.txt')
    assert_equal(pkg_writer.write_requires(), exp_out)
    assert_file_equal_string(exp_out,
"""# Requirements file for test-1
#
# Use with:
#
#     pip install -r test-1.txt

foo
bar
""")
    pkg_writer = PkgWriter('another', '2.0', '3.4.1',
                           ['baf==1.0', 'whack<2.1'])
    exp_out = pjoin(pkg_writer.wheel_build_dir, 'another-2.0.txt')
    assert_equal(pkg_writer.write_requires(), exp_out)
    assert_file_equal_string(exp_out,
"""# Requirements file for another-2.0
#
# Use with:
#
#     pip install -r another-2.0.txt

baf==1.0
whack<2.1
""")


def test_write_post():
    # Test write_post function
    pkg_writer = PkgWriter('test', '1', '3.4.1', ['foo', 'bar'],
                           wheel_sdir = 'pkgs')
    with TemporaryDirectory() as tmpdir:
        exp_out = pjoin(tmpdir, 'postinstall')
        assert_equal(pkg_writer.write_post(tmpdir), exp_out)
        assert_file_equal_string(exp_out,
"""#!/usr/bin/env python
# Install into Python.org python
# vim ft:python
import sys
import os
from os.path import exists, dirname
from subprocess import check_call

# Find disk image files
package_path = os.environ.get('PACKAGE_PATH')
if package_path is None:
    sys.exit(10)
package_dir = dirname(package_path)
wheelhouse = package_dir + '/pkgs'
# Find Python.org Python
python_bin = '/Library/Frameworks/Python.framework/Versions/3.4/bin'
python_path = python_bin + '/python3.4'
if not exists(python_path):
    sys.exit(20)
# Install pip
check_call([python_path,
            wheelhouse + '/get-pip.py',
            '-f', wheelhouse,
            '--no-index'])
# Find pip
expected_pip = python_bin + '/pip3.4'
if not exists(expected_pip):
    sys.exit(30)
check_call([expected_pip, 'install',
            '--no-index', '--upgrade',
            '--find-links', wheelhouse,
            '-r', wheelhouse + '/test-1.txt'])""")


def test_chatty_names():
    # Test existing chatty names property
    pkg_writer = PkgWriter('test', '1', '3.4.1', ['foo', 'bar'])
    assert_equal(pkg_writer.existing_chatty_names,
                 ('welcome.html', 'license.html'))


def test_pkg_name_pyv():
    # Test identifier
    pkg_writer = PkgWriter('test', '1.0', '3.4.1', ['foo', 'bar'])
    assert_equal(pkg_writer.pkg_name_pyv, 'test-py34')
    pkg_writer = PkgWriter('another', '2.0', '3.3.5', ['foo', 'bar'])
    assert_equal(pkg_writer.pkg_name_pyv, 'another-py33')


def test_pkg_name_pyv_version():
    # Test identifier
    pkg_writer = PkgWriter('test', '1.0', '3.4.1', ['foo', 'bar'])
    assert_equal(pkg_writer.pkg_name_pyv_version, 'test-py34-1.0')
    pkg_writer = PkgWriter('another', '2.0', '3.3.5', ['foo', 'bar'])
    assert_equal(pkg_writer.pkg_name_pyv_version, 'another-py33-2.0')


def test_identifier():
    # Test identifier
    pkg_writer = PkgWriter('test', '1.0', '3.4.1', ['foo', 'bar'])
    assert_equal(pkg_writer.identifier, 'com.github.MacPython.test-py34')
    pkg_writer = PkgWriter('another', '2.0', '3.3.5', ['foo', 'bar'],
                           pkg_id_root='org.dynevor.paul')
    assert_equal(pkg_writer.identifier,
                 'org.dynevor.paul.another-py33')


def test_dmg_no_clobber():
    # Test IOError if dmg exists
    pkg_writer = PkgWriter('test', '1.0', '3.4.1', ['foo', 'bar'])
    with TemporaryDirectory() as tmpdir:
        exp_name = pjoin(tmpdir, 'test-py34-1.0.dmg')
        with open(exp_name, 'wt') as fobj:
            fobj.write('My essential data')
        assert_raises(IOError, pkg_writer.write_dmg, tmpdir)



def test_template_override():
    # Check that we can override a template
    original_fname = pjoin(TEMPLATE_PATH, 'requirements.txt')
    with TemporaryDirectory() as tmpdir:
        new_fname = pjoin(tmpdir, 'requirements.txt')
        with open(new_fname, 'wt') as fobj:
            fobj.write('Nothing much')
        tpl = get_template('requirements.txt')
        assert_equal(tpl.filename, original_fname)
        insert_template_path(tmpdir)
        try:
            tpl = get_template('requirements.txt')
            assert_equal(tpl.filename, new_fname)
            assert_file_equal_string(new_fname, tpl.render())
        finally:
            pop_template_path()
        tpl = get_template('requirements.txt')
        assert_equal(tpl.filename, original_fname)


def test_get_template():
    tpl_fname = pjoin(TEMPLATE_PATH, 'requirements.txt')
    assert_equal(get_template('requirements.txt').filename, tpl_fname)
    assert_equal(get_template('unlikely_name.foo'), None)


wheels2dmg/tests/test_wheels2dmg_cmd.py
""" Testing wheels2dmg_cmd module
"""

from nose.tools import (assert_true, assert_false, assert_raises,
                        assert_equal, assert_not_equal)

from .scriptrunner import ScriptRunner

run_cmd = ScriptRunner().run_command


def test_exit_codes():
    # Test invalid calls to main
    # Too few argumets
    code, stdout, stderr = run_cmd(['wheels2dmg'], check_code=False)
    assert_equal(code, 2)
    code, stdout, stderr = run_cmd(['wheels2dmg', 'mypackage'],
                                   check_code=False)
    assert_equal(code, 2)
    # Test non-zero exit code for no requirements
    code, stdout, stderr = run_cmd(['wheels2dmg', 'mypackage', '1'],
                                   check_code=False)
    assert_equal(code, 12)


wheels2dmg/pkgbuilders.py
""" Utilities for building pkg installer and disk image from wheels
"""
from __future__ import division, print_function

import os
from os.path import (exists, join as pjoin, abspath, expanduser, dirname,
                     realpath)
import shutil
from subprocess import check_call
try:
    from urllib2 import urlopen # Python 2
    from urlparse import urlparse
except ImportError:
    from urllib.request import urlopen # Python 3
    from urllib.parse import urlparse
from tempfile import mkdtemp
import re
from glob import glob
from warnings import catch_warnings, simplefilter

from jinja2 import Environment, FileSystemLoader, TemplateNotFound

from delocate.delocating import delocate_wheel
from delocate.wheeltools import add_platforms

from .piputils import (make_pip_parser, recon_pip_args, get_requirements,
                       get_req_strings)

JINJA_LOADER = FileSystemLoader(pjoin(dirname(__file__), 'templates'))
JINJA_ENV = Environment(loader=JINJA_LOADER, trim_blocks=True)

# Installed location of Python.org Python
PY_ORG_BASE='/Library/Frameworks/Python.framework/Versions'

# Canonical get-pip.py URL
GET_PIP_URL = 'https://bootstrap.pypa.io/get-pip.py'
PKG_ID_ROOT = 'com.github.MacPython'

# Full Python version checker
PY_VERSION_RE = re.compile(r'\d\.\d\.\d+')


def get_get_pip(get_pip_url, out_dir):
    """ Get ``get-pip.py`` from file or URL `get_pip_url`, write to `out_dir`

    Parameters
    ----------
    get_pip_url : str
        File path or URL pointing to ``get-pip.py`` installer.  File path can
        contain ``~`` for home directory.
    out_dir : str
        Directory to which to write copy of ``get-pip.py``

    Returns
    -------
    get_pip_path : str
        Path to written ``get-pip.py``
    """
    parsed = urlparse(get_pip_url)
    get_pip_path = pjoin(out_dir, 'get-pip.py')
    if parsed.scheme == '': # File path
        gpp_path = expanduser(get_pip_url)
        shutil.copyfile(gpp_path, get_pip_path)
    else: # URL
        url_obj = urlopen(get_pip_url)
        with open(get_pip_path, 'wb') as fobj:
            fobj.write(url_obj.read())
    return get_pip_path


def get_python_path(pyv_m_m):
    """ Return path to Python.org python or raise error

    Parameters
    ----------
    pyv_m_m : str
        Python version in major.minor format (e.g. "2.7")

    Returns
    -------
    python_path : str
        Path to Python.org Python executable
    """
    python_path = '{0}/{1}/bin/python{1}'.format(PY_ORG_BASE, pyv_m_m, pyv_m_m)
    if not exists(python_path):
        raise RuntimeError('Need to install Python.org python version ' +
                           pyv_m_m)
    return python_path


def upgrade_pip(get_pip_path, pyv_m_m, pip_params):
    """ Upgrade pip with ``git-pip.py`` script, install ``wheel``

    Installs pip for Python.org Python if not present. Upgrades if necessary.
    Install ``wheel`` if necessary.

    Parameters
    ----------
    get_pip_path : str
        Path to local copy of ``get-pip.py``
    pyv_m_m : str
        Python version in major.minor format (e.g. "2.7")
    pip_params : sequence
        Parameters to pass to pip when installing

    Returns
    -------
    pip_exe : str
        Path to ``pip`` executable
    """
    python_path = get_python_path(pyv_m_m)
    # Upgrade pip
    check_call([python_path, get_pip_path] + pip_params)
    pip_exe = '{0}/{1}/bin/pip{1}'.format(PY_ORG_BASE, pyv_m_m, pyv_m_m)
    if not exists(pip_exe):
        raise RuntimeError('Expected to find pip at {0}, but not so'.format(
            pip_exe))
    # Install wheel
    check_call([pip_exe, 'install', '--upgrade'] + pip_params + ['wheel'])
    return pip_exe


def _safe_mkdirs(path):
    if not exists(path):
        os.makedirs(path)
    return path


class PkgWriter(object):
    """ Class to prepare and write DMG installer
    """
    py_org_base = PY_ORG_BASE
    pip_parser = make_pip_parser()
    chatty_names = ('welcome.html', 'readme.html', 'license.html')

    def __init__(self,
                 pkg_name,
                 pkg_version,
                 full_py_version,
                 pip_params,
                 get_pip_url = None,
                 dmg_build_dir = None,
                 scratch_dir = None,
                 pkg_id_root = None,
                 wheel_sdir = 'wheels',
                 wheel_component_name = 'wheel-installer',
                 delocate_wheels = True
                ):
        """ Initialize PkgWriter class

        Parameters
        ----------
        pkg_name : str
            The name of the package / installer
        pkg_version : str
            Version of the package / installer
        full_py_version : str
            Python version in form major.minor.extra - e.g. "3.4.1"
        pip_params : sequence
            Parameters to pip "wheel" command giving requirements and any index
            finding parameters.
        get_pip_url : None or str, optional
            URL or local path to ``get-pip.py`` command. We translate None to a
            the canonical pip URL
        dmg_build_dir : None or str, optional
            Directory in which to compile contents of disk image. Created if it
            does not exist.  If None, use temporary directory.
        scratch_dir : None or str, optional
            Directory in which to write files to construct installer packages.
            Created if it does not exist.  If None, use temporary directory.
        pkg_id_root : None or str, optional
            Root of 'identifier' used to identify package, in database of
            package receipts (given by ``pkgutil --pkgs``). The installer does
            not itself generate a package receipt, (see Notes).
        wheel_sdir : str, optional
            Name of wheelhouse directory on disk image
        wheel_component_name : str, optional
            Name of wheel component package within product archive .pkg file.
        delocate_wheels : bool, optonal
            If True, run ``delocate_wheel`` on all wheels in wheelhouse, to
            detect and maybe fix wheels built as part of the ``pip wheel``
            procedure to compile the wheelhouse.

        Notes
        -----
        On receipts: `pkg_id_root` gives the root receipt indentifier for the
        package, and the full identifier is given by the ``identifier``
        property below.  The identifier is the name of the "receipt" that goes
        into the database of installed package read and summarized by ``pkgutil
        --pkgs``. This installer does not write a receipt itself, because it is
        a "scripts only" installer (see ``man pkgbuild``), and therefore lacks
        a ``Bom`` (bill-of-materials) file in the wheel-installing compoenent
        package.  The only use of the identifier is so OSX doesn't confuse this
        package with another that might already be installed.
        """
        self.do_init()
        self.pkg_name = pkg_name
        self.pkg_version = pkg_version
        if not PY_VERSION_RE.match(full_py_version):
            raise ValueError('Need full Python version of form '
                             'major.minor.extra - e.g. "3.4.1"')
        self.full_py_version = full_py_version
        self.pip_params = pip_params
        self.get_pip_url = GET_PIP_URL if get_pip_url is None else get_pip_url
        self.dmg_build_dir = self._working_dir(dmg_build_dir)
        self.scratch_dir = self._working_dir(scratch_dir)
        self.pkg_id_root = PKG_ID_ROOT if pkg_id_root is None else pkg_id_root
        self.wheel_sdir = wheel_sdir
        self.wheel_component_name = wheel_component_name
        self.delocate_wheels = delocate_wheels

    def do_init(self):
        """ Extra initialization for object
        """
        self._to_delete = []

    def _working_dir(self, work_dir):
        """ Make working directory `work_dir`, return absolute path

        Parameters
        ----------
        work_dir : None or str
            If str, directory to create if it doesn't exist. If None, make a
            temporary directory and return that, noting that we have to delete
            when we've finished.

        Returns
        -------
        abs_work_dir : str
            Absolute path to working directory
        """
        if work_dir is None:
            work_dir  = mkdtemp()
            self._to_delete.append(work_dir)
        else:
            _safe_mkdirs(work_dir)
        return abspath(work_dir)

    def __del__(self):
        for pth in self._to_delete:
            shutil.rmtree(pth)

    # Versions of the python version string
    @property
    def pyv_m_m_e(self):
        """ Full version, e.g "3.4.1" """
        return self.full_py_version

    @property
    def pyv_m_m(self):
        """ Major.minor version, e.g "3.4" """
        return self.full_py_version[:3]

    @property
    def pyv_mm(self):
        """ Major,minor version with no dot e.g "34" """
        return self.full_py_version.replace('.', '')[:2]

    @property
    def pyv_m(self):
        """ Major version e.g "3" """
        return self.full_py_version[0]

    @property
    def pkg_name_version(self):
        return '{0}-{1}'.format(self.pkg_name, self.pkg_version)

    @property
    def pkg_name_pyv(self):
        return '{0}-py{1}'.format(
            self.pkg_name,
            self.pyv_mm)

    @property
    def pkg_name_pyv_version(self):
        return '{0}-{1}'.format(
            self.pkg_name_pyv,
            self.pkg_version)

    @property
    def wheel_build_dir(self):
        return pjoin(self.dmg_build_dir, self.wheel_sdir)

    @property
    def existing_chatty_names(self):
        return tuple(name for name in self.chatty_names
                     if not get_template(name) is None)

    @property
    def identifier(self):
        return '{0}.{1}'.format(self.pkg_id_root, self.pkg_name_pyv)

    def get_requirement_strings(self, extras=True, versions=True):
        """ Return list of requirement strings for requirements in `self`

        Parameters
        ----------
        extras : bool, optional
            If True, append extras specifications to requirement name
        versions : bool, optional
            If True, append version specifications to requirement name

        Returns
        -------
        req_strings : list
            list of string corresponding to the requirements `self`
        """
        args = self.pip_parser.parse_args(self.pip_params)
        req_set = get_requirements(args.req_specs, args.requirement)
        return get_req_strings(req_set, extras, versions)

    def get_wheels(self):
        """ Upgrade pip and get wheels for this install
        """
        wheelhouse = _safe_mkdirs(self.wheel_build_dir)
        # Get get-pip.py
        get_pip_path = get_get_pip(self.get_pip_url, wheelhouse)
        # Get pip arguments
        pip_args = self.pip_parser.parse_args(self.pip_params)
        req_params, fetch_params = recon_pip_args(pip_args)
        # Find or install pip, install wheel, for given Python.org Python
        pip_exe = upgrade_pip(get_pip_path, self.pyv_m_m, fetch_params)
        # Fetch the wheels we need
        check_call([pip_exe, 'wheel',
                    '-w', wheelhouse,
                    'pip', 'setuptools'] +
                   req_params + fetch_params)

    def process_wheels(self):
        """ Delocate built wheels, check archs, add platform tags
        """
        for wheel in glob(pjoin(self.wheel_build_dir, '*.whl')):
            if not '-macosx_10_6_intel' in wheel: # Pure wheel
                continue
            if self.delocate_wheels:
                with catch_warnings():
                    simplefilter('ignore')
                    delocate_wheel(wheel, require_archs='intel')
            new_wheel = add_platforms(
                wheel,
                ('macosx_10_9_intel', 'macosx_10_9_x86_64'),
                clobber=True)
            # returned new_wheel is None or absolute path
            if new_wheel and realpath(new_wheel) != realpath(wheel):
                os.unlink(wheel)

    def write_requires(self):
        """ Write a pip requirements file with given requirements

        Returns
        -------
        requires_fname : str
            Filename of written requirements file
        """
        requires_fname = pjoin(_safe_mkdirs(self.wheel_build_dir),
                               self.pkg_name_version + '.txt')
        template = get_template('requirements.txt')
        with open(requires_fname, 'wt') as fobj:
            fobj.write(template.render(info = self))
        return requires_fname

    def write_wheelhouse(self):
        """ Write wheels, requirements into wheelhouse directory
        """
        self.get_wheels()
        self.process_wheels()
        self.write_requires()

    def write_post(self, out_dir):
        """ Write ``postinstall`` file

        Parameters
        ----------
        out_dir : str
            Directory to which to write file

        Returns
        -------
        post_fname : str
            Filename of written file
        """
        post_fname = pjoin(out_dir, 'postinstall')
        template = get_template('postinstall')
        with open(post_fname, 'wt') as fobj:
            fobj.write(template.render(info = self))
        check_call(['chmod', 'a+x', post_fname])
        return post_fname

    def write_webloc(self):
        """ Write webloc link to Python installer
        """
        template = get_template('first_install_python.webloc')
        froot = 'Install Python {0}.webloc'.format(self.pyv_m_m)
        webloc_fname = pjoin(self.dmg_build_dir, froot)
        with open(webloc_fname, 'wt') as fobj:
            fobj.write(template.render(info = self))
        check_call(['SetFile', '-a', 'E', webloc_fname])
        return webloc_fname

    def write_readme(self):
        """ Write README.txt file """
        template = get_template('README.txt')
        readme_fname = pjoin(self.dmg_build_dir, 'README.txt')
        with open(readme_fname, 'wt') as fobj:
            fobj.write(template.render(info = self))
        return readme_fname

    def write_component_pkg(self):
        """ Write component package to install wheels

        Returns
        -------
        component_pkg_fname : str
            Filename of written compoent package
        """
        pkg_fname = pjoin(self.scratch_dir, self.wheel_component_name + '.pkg')
        scripts = pjoin(self.scratch_dir, 'scripts')
        _safe_mkdirs(scripts)
        self.write_post(scripts)
        check_call(['pkgbuild',
                    '--nopayload',
                    '--scripts', scripts,
                    '--identifier', self.identifier,
                    '--version', self.pkg_version,
                    pkg_fname])
        return pkg_fname

    def write_distribution(self):
        """ Write Distribution XML for product archive
        """
        template = get_template('Distribution')
        out_fname = pjoin(self.scratch_dir, 'Distribution')
        with open(out_fname, 'wt') as fobj:
            fobj.write(template.render(info = self))
        return out_fname

    def write_resources(self):
        """ Write contents of resources directory

        Returns
        -------
        resources : str
            Path of resources directory
        """
        resources = pjoin(self.scratch_dir, 'resources')
        en_resources = pjoin(resources, 'en.lproj')
        _safe_mkdirs(en_resources)
        for name in self.existing_chatty_names:
            out_fname = pjoin(en_resources, name)
            with open(out_fname, 'wt') as fobj:
                fobj.write(get_template(name).render(info = self))
        return resources

    def write_product_archive(self):
        """ Write product archive ``.pkg`` file
        """
        distribution = self.write_distribution()
        self.write_component_pkg()
        resources = self.write_resources()
        product_fname = pjoin(self.dmg_build_dir,
                              self.pkg_name_pyv_version + '.pkg')
        check_call(['productbuild',
                    '--distribution', distribution,
                    '--resources', resources,
                    '--package-path', self.scratch_dir,
                    product_fname])

    def write_dmg(self, out_dir, clobber=False):
        """ Write disk image ``.dmg`` file

        Parameters
        ----------
        out_dir : str
            Directory in which to write ``.dmg`` file
        clobber : bool, optional
            If True, overwrite existing file.  If False, raise IOError if file
            exists.
        """
        dmg_fname = pjoin(out_dir, self.pkg_name_pyv_version + '.dmg')
        if exists(dmg_fname):
            if not clobber:
                raise IOError(
                    '{0} exists, declining to overwrite'.format(dmg_fname))
            os.unlink(dmg_fname)
        self.write_webloc()
        self.write_readme()
        self.write_wheelhouse()
        self.write_product_archive()
        check_call(['hdiutil', 'create',
                    '-srcfolder', self.dmg_build_dir,
                    '-volname', self.pkg_name_pyv_version,
                    dmg_fname])
        return dmg_fname


def insert_template_path(path):
    """ Insert new path into jinja environment global

    Parameters
    ----------
    path : str
        Path to insert at beginning of template search order
    """
    JINJA_ENV.cache.clear()
    JINJA_LOADER.searchpath.insert(0, path)


def pop_template_path():
    """ Remove first path from jinja environment global

    Usually this is after using :func:`insert_templat_path`
    """
    JINJA_LOADER.searchpath.pop(0)
    JINJA_ENV.cache.clear()


def get_template(*args, **kwargs):
    """ Get template from jinja environment global

    Return None if template not found
    """
    try:
        return JINJA_ENV.get_template(*args, **kwargs)
    except TemplateNotFound:
        return None
